skimage/morphology/footprints.py

```diff
@@ -48,25 +173,71 @@ def rectangle(nrows, ncols, dtype=np.uint8):
 
     Other Parameters
     ----------------
-    dtype : data-type
+    dtype : data-type, optional
         The data type of the footprint.
+    decomposition : {None, 'separable', 'sequence'}, optional
+        If None, a single array is returned. For 'sequence', a tuple of smaller
+        footprints is returned. Applying this series of smaller footprints will
+        given an identical result to a single, larger footprint, but often with
+        better computational performance. See Notes for more details.
+        With 'separable', this function uses separable 1D footprints for each
+        axis. Whether 'seqeunce' or 'separable' is computationally faster may
+        be architecture-dependent.
 
     Returns
     -------
-    footprint : ndarray
+    footprint : ndarray or tuple
         A footprint consisting only of ones, i.e. every pixel belongs to the
-        neighborhood.
+        neighborhood. When `decomposition` is None, this is just a
+        numpy.ndarray. Otherwise, this will be a tuple whose length is equal to
+        the number of unique structuring elements to apply (see Notes for more
+        detail)
 
     Notes
     -----
+    When `decomposition` is not None, each element of the `footprint`
+    tuple is a 2-tuple of the form ``(ndarray, num_iter)`` that specifies a
+    footprint array and the number of iterations it is to be applied.
+
+    For binary morphology, using ``decomposition='sequence'``
+    was observed to give better performance, with the magnitude of the
+    performance increase rapidly increasing with footprint size. For grayscale
+    morphology with rectangular footprints, it is recommended to use
+    ``decomposition=None`` since the internal SciPy functions that are called
+    already have a fast implementation based on separable 1D sliding windows.
+
     - The use of ``width`` and ``height`` has been deprecated in
       version 0.18.0. Use ``nrows`` and ``ncols`` instead.
     """
-
-    return np.ones((nrows, ncols), dtype=dtype)
-
-
-def diamond(radius, dtype=np.uint8):
+    if decomposition is None:  # TODO: check optimal width setting here
+        return np.ones((nrows, ncols), dtype=dtype)
+
+    even_rows = nrows % 2 == 0
+    even_cols = ncols % 2 == 0
+    if decomposition == 'separable' or even_rows or even_cols:
+        sequence = [(np.ones((nrows, 1), dtype=dtype), 1),
+                    (np.ones((1, ncols), dtype=dtype), 1)]
+    elif decomposition == 'sequence':
+        # this branch only support odd nrows, ncols
+        sq_size = 3
+        sq_reps = _decompose_size(min(nrows, ncols), sq_size)
+        sequence = [(np.ones((3, 3), dtype=dtype), sq_reps)]
+        if nrows > ncols:
+            nextra = nrows - ncols
+            sequence.append(
+                (np.ones((nextra + 1, 1), dtype=dtype), 1)
+            )
+        elif ncols > nrows:
```

Comment:
what happens when `ncols == nrows`?

Reply:
The sequence defined before the `if nrows > ncols` is the full sequence in that case (the footprint is square, so we don't need to append additional anisotropic element(s) to stretch things along a particular dimension).